physionet-django/user/views.py

```diff
@@ -459,30 +469,26 @@ def auth_orcid(request):
     """
 
     client_id = settings.ORCID_CLIENT_ID
-    client_secret = settings.ORCID_CLIENT_SECRET
     redirect_uri = settings.ORCID_REDIRECT_URI
     scope = list(settings.ORCID_SCOPE.split(","))
     oauth = OAuth2Session(client_id, redirect_uri=redirect_uri,
                           scope=scope)
     params = request.GET.copy()
     code = params['code']
 
-    try:
-        token = oauth.fetch_token(settings.ORCID_TOKEN_URL, code=code,
-                                  include_client_id=True, client_secret=client_secret)
-        try:
-            validators.validate_orcid_token(token['access_token'])
-            token_valid = True
-        except ValidationError:
-            messages.error(request, 'Validation Error: ORCID token validation failed.')
-            token_valid = False
-    except InvalidGrantError:
-        messages.error(request, 'Invalid Grant Error: authorization code may be expired or invalid.')
-        token_valid = False
+    token_valid, token = _fetch_and_validate_token(request, code, oauth)
 
     if token_valid:
-        orcid_profile, _ = Orcid.objects.get_or_create(user=request.user)
-        orcid_profile.orcid_id = token.get('orcid')
+        orcid_id = token.get('orcid')
+        orcid_profile = Orcid.objects.filter(orcid_id=orcid_id).first()
+        if orcid_profile and orcid_profile.user != request.user:
```

Comment:
My understanding here is that we:

1. Get the ORCID
2. Check if this ORCID is associated with a user
3. If the ORCID is not associated with a user, create a new user account.

This is quite problematic. Most (or at least many) users will have an ORCID but the ORCID will not be linked to their PhysioNet account. In these cases, the user will end up with duplicate accounts. We do not have a method for merging/linking PhysioNet accounts right now.

Reply:
You can see my answer [below](https://github.com/MIT-LCP/physionet-build/pull/2276#issuecomment-2548216926)